spyder_kernels/console/kernel.py

```diff
@@ -114,71 +166,25 @@ def get_var_properties(self):
                     'array_ndim': self._get_array_ndim(value)
                 }
 
-            return repr(properties)
+            return properties
         else:
-            return repr(None)
+            return None
 
-    def send_spyder_msg(self, spyder_msg_type, content=None, data=None):
-        """
-        Publish custom messages to the Spyder frontend.
-
-        Parameters
-        ----------
-
-        spyder_msg_type: str
-            The spyder message type
-        content: dict
-            The (JSONable) content of the message
-        data: any
-            Any object that is serializable by cloudpickle (should be most
-            things). Will arrive as cloudpickled bytes in `.buffers[0]`.
-        """
-        import cloudpickle
-
-        if content is None:
-            content = {}
-        content['spyder_msg_type'] = spyder_msg_type
-        msg = self.session.send(
-            self.iopub_socket,
-            'spyder_msg',
-            content=content,
-            buffers=[cloudpickle.dumps(data, protocol=PICKLE_PROTOCOL)],
-            parent=self._parent_header,
-        )
-        self.log.debug(msg)
+    def update_var_properties(self):
+        """Send updated var properties."""
+        properties = self.get_var_properties()
+        self.frontend_call().set_var_properties(properties)
 
     def get_value(self, name):
         """Get the value of a variable"""
         ns = self._get_current_namespace()
-        value = ns[name]
-        try:
-            self.send_spyder_msg('data', data=value)
-        except:
-            # * There is no need to inform users about
-            #   these errors.
-            # * value = None makes Spyder to ignore
-            #   petitions to display a value
-            self.send_spyder_msg('data', data=None)
         self._do_publish_pdb_state = False
+        return ns[name]
 
-    def set_value(self, name, value, PY2_frontend):
+    def set_value(self, name, value):
         """Set the value of a variable"""
-        import cloudpickle
         ns = self._get_reference_namespace(name)
-
-        # We send serialized values in a list of one element
-        # from Spyder to the kernel, to be able to send them
-        # at all in Python 2
-        svalue = value[0]
-
-        # We need to convert svalue to bytes if the frontend
-        # runs in Python 2 and the kernel runs in Python 3
-        if PY2_frontend and not PY2:
-            svalue = bytes(svalue, 'latin-1')
```

Comment:
How are you handling this case where the frontend runs in Python 2 and the kernel in Python 3?

It seems that you simply dismissed what we already had done before.

Reply:
Well the value could be anything, not just a string. the pickling is done in the comms. This function doesn't work with string anymore.

Reply:
This might be something that needs to be fixed in the comms instead.

Reply:
And as is it doesn't work for me as the encoding I get from cloudpickle.dumps in python 2 is `utf8` and not `latin-1`.

Reply:
This was not added here to handle strings. `svalue` was the serialized value of an object sent by Spyder to be set here in the kernel, after making changes to it through one of our editors (e.g. the dataframe editor).

The problem we solved here is that if the frontend is running in Python 2 and the kernel in Python 3, we need to convert that serialized value to `latin-1` before trying to deserialize it. Else we won't get the right value.

Now, since you move all communications to use serialization (before it was a mix of serialization and passing strings), I think you need to do this to before deserializing anything that comes from frontend and fulfills this case.

And you should also apply this trick in the frontend, if that runs in Python 3 and the kernel in Python 2 (but that could be done in CommsBase, right?)

Reply:
What I mean is that there might be other problems in encoding:

In python 2:
```python
cloudpickle.dumps("eéè", 2)
'\x80\x02U\x05e\xc3\xa9\xc3\xa8q\x00.'
```

In python3:
```python
value = '\x80\x02U\x05e\xc3\xa9\xc3\xa8q\x00.'
bvalue = bytes(value, 'latin-1')
cloudpickle.loads(bvalue)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
cloudpickle.loads(bvalue, encoding='latin-1')
'eÃ©Ã¨'
cloudpickle.loads(bvalue, encoding='utf8')
'eéè'
```

Reply:
from https://docs.python.org/3/library/pickle.html#pickle.loads :

> 
    Optional keyword arguments are fix_imports, encoding and errors,
    which are used to control compatibility support for pickle stream 
    generated by Python 2.

Reply:
Ok, then this means we need to use the same encoding to serialize and deserialize on both sides.

Reply:
> Using encoding='latin1' is required for unpickling NumPy arrays and instances of datetime, date and time pickled by Python 2.

I'll do that. There might be encoding problems but I don't really have a choice. (Anyway encoding problems is the reason why python2 still exists)

Reply:
Well, if there are encoding problems, `None` should be sent over the wire.

And thanks for bringing a better understanding about this!

Reply:
Ok this is not a problem because I send the values in the 'buffer' part of the messages which is just data, not a python object such as `string` or `bytes`. BUT, the encoding issue is still present, so I will fix that.

Reply:
`svalue = bytes(svalue, 'latin-1')` is only iseful to cast a string object to a bytes object, but the buffer will be opened correctly for any combination (as it is just `memory`).

Reply:
> Ok, then this means we need to use the same encoding to serialize and deserialize on both sides.

We can't do that as we have no control of the encoding in python2. Basically the encoding will be: whatever the user decides, and the decoding will be 'latin-1', because otherwise numpy won't work. This means that anything that is encoded with utf-8 will be corrupted. (The previous behaviour was just crashing, so this is an improvment.) I don't think there is anything we can do. This kind of problem is exactly why python3 is a breaking update.

Reply:
> Well, if there are encoding problems, `None` should be sent over the wire.

We won't know if there are encoding problems. The behaviour in python2 is : encode with whatever, decode with watever, hope it works. The behaviour in python3 is: encode with utf8, decode with utf8. If a python 3 string is sent to python 2, it will be shown as an unicode string (Which might be surprising, but at least will be decoded correctly).

Reply:
I think pickled string are encoded as `latin-1` in Python 2. At least, that's what I found with a simple test in IPython:

```python
In [1]: import cloudpickle

In [2]: svalue = cloudpickle.dumps('ñ')

In [3]: svalue.decode()
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-3-ab1c9e779743> in <module>()
----> 1 svalue.decode()

UnicodeDecodeError: 'ascii' codec can't decode byte 0x80 in position 0: ordinal not in range(128)

In [4]: svalue.decode('latin-1')
Out[4]: u'\x80\x02U\x02\xc3\xb1q\x00.'
```

Reply:
- Python 2 strings don't have encoding, they are basically `bytes` from python3. The encoding depends on where the strings come from. (File, keyboard, or other)
 - Python 3 string are UTF-8 encoded data.
 - If you want to display a string in Python2, you have to know what the encoding is. If you don't, you are out of luck, and will end up displaying computer gibbrish if you have non ascii letters. This is a major problem with python2, and the reason why python3 is a breaking update.
 - In Python3, you know the encoding, so no problem

- Now pickle want to serialise a list of strings. In Python2 pickle can not know anything about encoding, and therefore saves the string as is.
 - In Python3, pickle must create strings as unicode strings, but it has no clue what the encoding of the strings are (the encoding of each string could be different.) Therefore there is a few options:
    - Open all the strings as `bytes`. This is the correct way but then the user has to decode every single string that was pickled.
    - Open all the strings as `ascii`. Works for all ascii strings but fails as soon as a non ascii charater is found. This is the default for pickle.
    - Choose a random encoding and accept that the non-ascii letters might be gibbrish. This is what we do here.

Reply:
It seems you understand better the differences between Python 2 and 3 string representations than me! So I defer to your judgment here.